src/main/kotlin/com/amazon/opendistroforelasticsearch/indexstatemanagement/step/rollover/AttemptRolloverStep.kt

```diff
@@ -0,0 +1,152 @@
+
+    private val logger = LogManager.getLogger(javaClass)
+    private var failed: Boolean = false
+    private var stepCompleted: Boolean = false
+    private var info: Map<String, Any>? = null
+
+    // TODO: Incorporate retries from config and consumed retries from metadata
+    override suspend fun execute() {
+        // If we have already rolled over this index then fail as we only allow an index to be rolled over once
+        if (managedIndexMetaData.rolledOver == true) {
+            failed = true
+            info = mapOf("message" to "This index has already been rolled over")
+            return
+        }
+
+        val alias = getAlias()
+
+        alias ?: return
+
+        val statsResponse = getIndexStats()
+
+        statsResponse ?: return
+
+        val indexCreationDate = Instant.ofEpochMilli(clusterService.state().metaData().index(managedIndexMetaData.index).creationDate)
+        val numDocs = statsResponse.primaries.docs.count
+        val indexSize = ByteSizeValue(statsResponse.primaries.docs.totalSizeInBytes)
+
+        if (config.evaluateConditions(indexCreationDate, numDocs, indexSize)) {
+            executeRollover(alias)
+        } else {
+            info = mapOf("message" to "Attempting to rollover")
+        }
+    }
+
+    private suspend fun executeRollover(alias: String) {
+        try {
+            val request = RolloverRequest(alias, null)
+            val response: RolloverResponse = client.admin().indices().suspendUntil { rolloverIndex(request, it) }
+
+            // Do not need to check for isRolledOver as we are not passing any conditions or dryrun=true
+            // which are the only two ways it comes back as false
+
+            // If response isAcknowledged it means the index was created and alias was added to new index
+            if (response.isAcknowledged) {
+                stepCompleted = true
+                info = mapOf("message" to "Rolled over index")
+            } else {
+                // If the alias update response is NOT acknowledged we will get back isAcknowledged=false
+                // This means the new index was created but we failed to swap the alias
+                failed = true
+                info = mapOf("message" to "New index created (${response.newIndex}), but failed to update alias")
+            }
+        } catch (error: Exception) {
+            failed = true
+            val mutableInfo = mutableMapOf("message" to "Failed to rollover index")
+            val errorMessage = error.message
+            if (errorMessage != null) mutableInfo.put("cause", errorMessage)
+            info = mutableInfo.toMap()
+        }
+    }
+
+    private fun getAlias(): String? {
+        val alias = clusterService.state().metaData().index(managedIndexMetaData.index).getRolloverAlias()
+
+        if (alias == null) {
+            failed = true
+            info = mapOf("message" to "There is no valid rollover_alias=$alias set on ${managedIndexMetaData.index}")
+        }
```

Comment:
I think this if block in here can be counter intuitive. From the function name it sounds like only thing this should do is `getAlias` but it has some business logic that needs to be handled outside of this function. So when I initially saw the 

```
val alias = getAlias()

alias ?: return
```

In `execute` I was unaware this is what was happen in here. I think it makes more sense to let `getAlais` to only get `alias` and let the error handling to be done in the `execute`. Other implementation that may make sense is `getAlias` throwing an exception if the `alias` is null.

Reply:
We update the info/failed different ways depending on errors, response status, etc. and it pollutes the execute method whereas I think we should try to keep execute a high level overview of what's happening in this step. I agree the naming is confusing `getAlias`, changing to a more verbose name for now until we can eventually come up with a better solution to the metadata and execute.